showcase/src/elm/UrlGenerator.elm

```diff
@@ -0,0 +1,513 @@
+
+                ExpectingEnd ->
+                    "ExpectingEnd at " ++ position
+
+                UnexpectedChar ->
+                    "UnexpectedChar at " ++ position
+
+                Problem str ->
+                    "ProblemString " ++ str ++ " at " ++ position
+
+                BadRepeat ->
+                    "BadRepeat at " ++ position
+    in
+    List.foldl (++) "" (List.map deadEndToString deadEnds)
+
+
+
+-- for example, "Routes.AlertComponent.BasicExample" -> "BasicExample AlertComponent"
+-- if parsing fails, it just return "bruh moment"
+
+
+getTitle : File -> String
+getTitle f =
+    let
+        splitName =
+            String.split "." f.moduleDeclaration.name
+                |> List.drop 1
+
+        -- remove "Routes"
+    in
+    case splitName of
+        [ component, name ] ->
+            name ++ " - " ++ component
+
+        _ ->
+            "bruh moment"
+
+
+
+-------------------
+-- MANIPULATORS
+-------------------
+
+
+changeModuleNameTo : String -> File -> File
+changeModuleNameTo newName file =
+    let
+        oldModuleDeclaration =
+            file.moduleDeclaration
+
+        newModuleDeclaration =
+            { oldModuleDeclaration | name = newName }
+    in
+    { file | moduleDeclaration = newModuleDeclaration }
+
+
+
+-- this only does one task so far
+-- if the name is 'example', then change it to 'main
+
+
+elliefyExports : File -> File
+elliefyExports file =
+    case file.moduleDeclaration.exposes of
+        [ "example" ] ->
+            let
+                newExposes =
+                    [ "main" ]
+
+                mDeclaration =
+                    file.moduleDeclaration
+
+                newMDeclaration =
+                    { mDeclaration | exposes = newExposes }
+
+                -- rename instances of "example" to "main"
+                newCodeBlocks =
+                    file.codeBlocks
+                        |> List.map
+                            (\block ->
+                                case block of
+                                    FunDef funData ->
+                                        if funData.funName == "example" then
+                                            FunDef
+                                                { funData | funName = "main" }
+
+                                        else
+                                            FunDef funData
+
+                                    typeDef ->
+                                        -- do not change type definitions
+                                        typeDef
+                            )
+            in
+            { file
+                | moduleDeclaration = newMDeclaration
+                , codeBlocks = newCodeBlocks
+            }
+
+        _ ->
+            -- do nothing - yet!!
+            file
+
+
+
+-------------------
+-- STRINGIFY
+-------------------
+
+
+toString : File -> String
+toString file =
```

Comment:
There are a few moving parts in this module, and in the context of maintainability, I think it would be nice to test more than just `fromSourceCode` which is currently the only exposed fn in this Module. 

There are 3 kinds of functions / code I see in this module:

- URL generation
- Anything to do with parsing and creating a `File`
- Modifying a `File` to adhere to the Ellie API

Thus, I think you could actually have a second module called `Utils.FileParser` which houses all the `File` related logic.

Here you would expose some abstraction over `Parser.run fileParser srcCode`.

```elm
-- something like this maybe?
parseFile : String -> Result ParseError File
parseFile =
    Parser.run fileParser
```

Then a test or 2 could be written for `parseFile`. 

--- 

As for the modification code, I'm not sure where I stand on whether that code should be moved to another module or not right now. Might be overkill at this moment. But if it gets more complex in the future, then it would be justified to move it into its own module and add tests for this code too.

Reply:
A separate module just for the file parsing would be nice, especially since the file parser is not a trivial function. I've separated the `UrlGenerator.elm` into `EllieLinks` and `FileParser`.

I think we can keep the modification code in the file parser code right now, but I can see it getting pretty complex. Right now, I don't think we need to move it. 

I didn't make much changes to the functions in the `FileParser.elm` file, so it still has the `elliefy` as the export. I'll get around to fixing this later, so I added a new action item to the PR.